### PR TITLE
Pioneer DDJ-SB mapping

### PR DESCRIPTION
**Pull request by jardiacaj**
I continued my work on the Pioneer DDJ-SB mapping. You can read the forum topic on this, including in page 3 a draft for the corresponding wiki page, here:

http://mixxx.org/forums/viewtopic.php?f=7&t=6886&sid=bab2d5720d8eb823a973cfca92cf5506

This mapping supports all the controller's features, including features blocked in Serato DJ Intro (which is bundled with the controller) and 4 deck control. I found no issues in the latest version of the mapping.

### CODE REVIEW

File: res/controllers/Pioneer-DDJ-SB-scripts.js [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+ */
+
+///////////////////////////////////////////////////////////////
+//                       USER OPTIONS                        //
+///////////////////////////////////////////////////////////////
+
+// If true the sync button blinks with the beat, if false led is lit when sync is enabled.
+PioneerDDJSB.blinkingSync = false;
+
+// It true, the vinyl button activates slip. Vinyl mode is then activated by using shift.
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: Here's a typo. Change "It" to "if".

File: res/controllers/Pioneer-DDJ-SB-scripts.js [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+    PioneerDDJSB.bindDeckControlConnections('[Channel3]', true);
+    PioneerDDJSB.bindDeckControlConnections('[Channel4]', true);
+};
+
+
+///////////////////////////////////////////////////////////////
+//                      CONTROL BINDING                      //
+///////////////////////////////////////////////////////////////
+
+PioneerDDJSB.bulkBindControlConnections = function (group, controlsToFunctions, remove) {
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: Good idea to make this its own function. I'll make use of that in my mapping.

File: res/controllers/Pioneer-DDJ-SB-scripts.js [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+    if (value) {
+        if (deckNumber <= 2) {
+            deckNumber += 2;
+        } else {
+            deckNumber -= 2;
+        }
+        PioneerDDJSB.deckSwitchTable[group] = '[Channel' + deckNumber + ']';
+        PioneerDDJSB.initDeck(PioneerDDJSB.deckSwitchTable[group], false);
+        PioneerDDJSB.setDeckSoftTakeover(PioneerDDJSB.deckSwitchTable[group], true);
+        PioneerDDJSB.setDeckSoftTakeover(PioneerDDJSB.deckSwitchTable[group]);
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: I'm confused about this. Why disconnect soft takeover then enable it again right away for the same deck? Is there really a need to disable soft takeover for any controls? Even if a different deck is active on the controller, I don't see the harm in leaving soft takeover enabled.

File: res/controllers/Pioneer-DDJ-SB-scripts.js [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+
+PioneerDDJSB.toggleEngineValue = function (group, control) {
+    engine.setValue(group, control, !(engine.getValue(group, control)));
+};
+
+
+///////////////////////////////////////////////////////////////
+//             TWO MESSAGE DECK INPUT HANDLERS               //
+///////////////////////////////////////////////////////////////
+
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: It would be great if you added a section to the MIDI Scripting wiki page about how to use high resolution MIDI.

File: res/controllers/Pioneer-DDJ-SB-scripts.js [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
+    if (!isFirstBinding) {
+        PioneerDDJSB.bindDeckControlConnections('[Channel' + disconnectDeck + ']', true);
+    }
+
+    PioneerDDJSB.bindDeckControlConnections(group, false);
+    PioneerDDJSB.nonPadLedControl(group, PioneerDDJSB.nonPadLeds.shiftKeyLock, disconnectDeck <= 2);
+    PioneerDDJSB.triggerVinylLed(PioneerDDJSB.channelGroups[group]);
+};
+
+PioneerDDJSB.toggleEngineValue = function (group, control) {
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: Could you move this to common-controller-scripts.js and use the "script" namespace so all scripts can take advantage of it?

File: res/controllers/Pioneer-DDJ-SB-scripts.js [hunk 6]
```diff
@@ -0,0 +1,888 @@
...
+    }
+};
+
+
+///////////////////////////////////////////////////////////////
+//                          JOGWHEELS                        //
+///////////////////////////////////////////////////////////////
+
+// Work out the jog-wheel change / delta
+PioneerDDJSB.getJogWheelDelta = function (value) {
```
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: This is so simple that I don't think it really needs to be a separate function.

**Comment by Be-ing** [comment]
Maybe it would be more appropriate to make shift+filter the gain knob rather than shift+effects knob. Gain knobs are usually in mixer sections right by EQ knobs.

**Comment by Be-ing** [comment]
It would be nice if you document how this mapping differs from the Serato DJ Intro and Serato DJ mappings.

File: res/controllers/Pioneer-DDJ-SB-scripts.js (hunk 5 above)
Comment (sblaisot) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: IMHO implementing this in common-controller-scripts.js is a good thing but needs to be in a separate PR.

File: res/controllers/Pioneer-DDJ-SB-scripts.js (hunk 1 above)
Comment (jardiacaj) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: Fixed.

File: res/controllers/Pioneer-DDJ-SB-scripts.js (hunk 2 above)
Comment (jardiacaj) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: I moved it for now.

File: res/controllers/Pioneer-DDJ-SB-scripts.js (hunk 3 above)
Comment (jardiacaj) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: It is the only way I got soft takeover to work. If this is not done like this, soft takeover does not work properly. The first time I switch decks it works, but when switching back the control values change abruptly as if soft takeover was disabled. I don't know if this is the desired behavior. For me it seems like a bug in Mixxx.

The effects need even a worse looking workaround, probably because of the same issue (but turning soft takeover off and on again did not make the trick there for some reason).

**Comment by jardiacaj** [comment]
I also think that making shift+filter the gain knob would be a good idea.

I will be on vacation for the next two weeks and probably won't have an Internet connection at all. I'll gladly continue working on this afterwards, just don't be surprised if for this couple of weeks I'm absent.

File: res/controllers/Pioneer-DDJ-SB-scripts.js (hunk 4 above)
Comment (jardiacaj) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: I can do that, it will have to be after my vacation :)

File: res/controllers/Pioneer-DDJ-SB-scripts.js (hunk 3 above)
Comment (Be-ing) [inline res/controllers/Pioneer-DDJ-SB-scripts.js]: If you figure out what the issue is more precisely, please open a bug report on Launchpad.

**Comment by Be-ing** [comment]
I'd prefer to not merge this until the gain control is moved to shift+filter. Other than that and the long function name I commented on, it looks good.

**Comment by bikedude880** [comment]
You can't do shift + filter according to the midi specs. Shift effect is the only logical way to integrate it

**Comment by Be-ing** [comment]
Huh? Why wouldn't that work?

**Comment by bikedude880** [comment]
I could be mistaken, but in my research for a 1.12 mapping, the sb midi spec v1.0 doesn't list any shift values for the effects knobs

**Comment by Be-ing** [comment]
Oh I didn't realize the controller firmware sends different MIDI values when shift is pressed. Does the shift button send note on/off signals? If so, the script could keep track of whether it is pressed and check that variable in the filter knob handling function.

See http://mixxx.org/wiki/doku.php/midi_scripting#modifier_shift_buttons

**Comment by bikedude880** [comment]
It sends midi note 9n ( n=0-4)  3f.  This shift button modifies values on the board internally as well (performance pads are the primary use)

**Comment by bikedude880** [comment]
Sorry, n=0 or 1 based on the deck

**Comment by bikedude880** [comment]
Been playing around with this mapping, noticed an issue where setting hot cues on deck 3 or 4 will set both cute points for both decks

**Comment by Be-ing** [comment]
What exactly do you mean? Setting a hotcue on deck 3 will also set one on deck 1, or it will also set one on deck 4? The hotCueButtons function uses the same method for choosing the deck as every other function handling a deck-specific control, so I don't know why that's happening...

**Comment by bikedude880** [comment]
I'll verify the exact details today after work and ensure I don't have anything else interfering.  It should be deck 3+4 getting duplicated hotcue events

**Comment by bikedude880** [comment]
Hot cue works exactly like it should, failed to notice I loaded the same track in multiple decks.

**Comment by jardiacaj** [comment]
Now shift+filter controls pregain. Also the "channel fader start" functionality has been implemented.

Shift+fx knob also allows to control pregain. I can currently not think of a better use for this. Any ideas?

Otherwise I'm quite satisfied with the mapping for now.

**Comment by Be-ing** [comment]
Thank you for documenting this on the wiki. The "channel fader start" functionality sounds fun. :)

> Shift+fx knob also allows to control pregain. I can currently not think of a better use for this. Any ideas?

I think mapping it to the effect chain's superknob would make sense.

The method enabling effects on decks 3 + 4 seems quite clunky to me. How about changing the effect enable buttons to apply the effect unit to decks 3 + 4 rather than 1 + 2 if that side of the controller is on deck 3 or 4?

**Comment by sblaisot** [comment]
Please don't forget to embed your new controller preset files in windows installer
https://github.com/mixxxdj/mixxx/blob/1.12/build/nsis/Mixxx.nsi#L343

**Comment by bikedude880** [comment]
Love "channel fader start" functionality.  I agree on the effect enable/superknob ideas.

**Comment by jardiacaj** [comment]
I added the files to the Windows installer (but did not test it).

I also made shift + FX knob control super1 instead of gain.

I also changed how effects are enabled/disabled. Now the center button (number 2) toggles the effect for the headphones, and the left and right buttons (1 and 3) toggle the effect for the left and right decks respectively, such that they always control the effects for the active decks (1 or 3 and 2 or 4).

**Comment by Be-ing** [comment]
Nice. I think this is ready to merge. :)

**Comment by Be-ing** [comment]
Please update the description on the wiki to reflect these changes.

**Comment by sblaisot** [comment]
LGTM

**Comment by Be-ing** [comment]
Now that you have gone through the process of making this mapping, I'd appreciate your input on the thread on the mailing list about improving the mapping format: http://sourceforge.net/p/mixxx/mailman/mixxx-devel/thread/5588554B.40904%40gmx.com/#msg34230237

**Comment by bikedude880** [comment]
Only left shift works to trigger things like gain, works well on both sides for super knobs.

Deck effects enable is tricky to master since the buttons change their settings when changing decks. Once it is understood how it works it's okay.

**Comment by jardiacaj** [comment]
I would like to fix the gain bug. How should this be done? Would it be a new PR?

**Comment by Be-ing** [comment]
Yes, now that this PR has been merged, if you want to change something, open a new PR.
